### DISCUSSION TITLE
Consider recommending `cargo +nightly docs-rs --open` for local documentation builds

### DISCUSSION THREAD

**Discussion by dtolnay** [tokio-rs/tokio#6469] [General] [answered]
> [!NOTE]
> Full disclosure: I wrote this crate. https://github.com/dtolnay/cargo-docs-rs

I noticed this in the contribution guide:

https://github.com/tokio-rs/tokio/blob/01ed7b55f753b104ec03627e1b6238287cc09c4f/CONTRIBUTING.md#L155-L169

For my own projects, iterating on docs locally and also checking the documentation build in CI, I have a `cargo docs-rs` subcommand that loads the "package.metadata.docs.rs" section of Cargo.toml, and then delegates to a `cargo rustdoc` command with all those options applied in the same way that docs.rs's builds.

https://github.com/tokio-rs/tokio/blob/01ed7b55f753b104ec03627e1b6238287cc09c4f/tokio/Cargo.toml#L157-L163

The output of `cargo docs-rs --verbose` shows what this turns into:

```console
$ cargo docs-rs --verbose
     Running `cargo rustdoc -Zunstable-options -Zrustdoc-map -Zhost-config -Ztarget-applies-to-host
            --lib --all-features --target=x86_64-unknown-linux-gnu
            --config='build.rustflags=["--cfg", "tokio_unstable", "--cfg", "tokio_taskdump"]'
            --config='host.rustflags=["--cfg", "tokio_unstable", "--cfg", "tokio_taskdump"]'
            --config='build.rustdocflags=["-Zunstable-options", "--cfg=docsrs", "--cfg", "docsrs", "--cfg", "tokio_unstable", "--cfg", "tokio_taskdump", "--extern-html-root-takes-precedence"]'
            --config='doc.extern-map.registries.crates-io="https://docs.rs"'
            --verbose`
```

This matches the rustdoc command in tokio's docs.rs build log <https://docs.rs/crate/tokio/1.37.0/builds/1172427> aside from some docs.rs-specific details (`--emit=invocation-specific`, `--resource-suffix`, `--static-root-path`, `-j6`).

**Comment by Darksonn** [comment, marked as answer]
Seems reasonable enough to me. I would still want the existing command in the documentation for people who don't want to install the tool, but making this the first recommendation sounds good to me.

We currently have two commands in the docs: one with and one without unstable features. But we should probably drop the one without unstable features since it isn't how docsrs will build it, and it emits warnings about broken links.

Do you want to submit a PR?

**Reply by dtolnay** [reply]
Sure thing: https://github.com/tokio-rs/tokio/pull/6475. Thanks!
